presto-hive/src/main/java/io/prestosql/plugin/hive/util/FieldSetterFactory.java

```diff
@@ -361,24 +375,60 @@ public void setField(Block block, int position)
         }
     }
 
-    private static class TimestampFieldSetter
+    private abstract static class TimestampFieldSetter
             extends FieldSetter
     {
-        private final DateTimeZone timeZone;
-        private final TimestampWritableV2 value = new TimestampWritableV2();
+        protected final DateTimeZone timeZone;
+        protected final TimestampWritableV2 value = new TimestampWritableV2();
 
         public TimestampFieldSetter(SettableStructObjectInspector rowInspector, Object row, StructField field, DateTimeZone timeZone)
         {
             super(rowInspector, row, field);
             this.timeZone = requireNonNull(timeZone, "timeZone is null");
         }
 
+        protected Timestamp getTimestamp(long micros, IntSupplier picosSupplier)
+        {
+            long epochMillis = floorDiv(micros, MICROSECONDS_PER_MILLISECOND);
+            epochMillis = timeZone.convertLocalToUTC(epochMillis, false);
+            int nanosFromMicros = floorMod(micros, MICROSECONDS_PER_SECOND) * NANOSECONDS_PER_MICROSECOND;
```

Comment:
up to millis precision is covered by `epochMillis` parameter, so
`mod MICROSECONDS_PER_SECOND` should probably be `mod MICROSECONDS_PER_MILLISECOND`

Reply:
It's definitely no clear what's happening here, but the `Timestamp` we are returning is constructed from epoch millis and *nanos of second* (since it's using a `java.time.LocalDateTime` under the covers).  Pretty intuitive API, right?  I'll try to make this explicit.